src/backend/distributed/executor/distributed_intermediate_results.c

```diff
@@ -318,3 +379,227 @@ ExecuteSelectTasksIntoTupleStore(List *taskList, TupleDesc resultDescriptor)
 
 	return resultStore;
 }
+
+
+/*
+ * ColocateFragmentsWithRelation moves the fragments in the cluster so they are
+ * colocated with the shards of target relation. These transfers are done by
+ * calls to fetch_intermediate_results() between nodes.
+ *
+ * returnValue[shardIndex] is list of result Ids that are colocated with
+ * targetRelation->sortedShardIntervalArray[shardIndex] after fetch tasks are
+ * done.
+ */
+static List **
+ColocateFragmentsWithRelation(List *fragmentList, DistTableCacheEntry *targetRelation)
+{
+	List *fragmentListTransfers = ColocationTransfers(fragmentList, targetRelation);
+	List *fragmentTransferTaskList = FragmentTransferTaskList(fragmentListTransfers);
+
+	ExecuteFetchTaskList(fragmentTransferTaskList);
+
+	int shardCount = targetRelation->shardIntervalArrayLength;
+	List **shardResultIdList = palloc0(shardCount * sizeof(List *));
+
+	ListCell *fragmentCell = NULL;
+	foreach(fragmentCell, fragmentList)
+	{
+		DistributedResultFragment *sourceFragment = lfirst(fragmentCell);
+		int shardIndex = sourceFragment->targetShardIndex;
+
+		shardResultIdList[shardIndex] = lappend(shardResultIdList[shardIndex],
+												sourceFragment->resultId);
+	}
+
+	return shardResultIdList;
+}
+
+
+/*
+ * ColocationTransfers returns a list of transfers to colocate given fragments with
+ * shards of the target relation. These transfers also take into account replicated
+ * target relations. This prunes away transfers with same source and target
+ */
+static List *
+ColocationTransfers(List *fragmentList, DistTableCacheEntry *targetRelation)
+{
+	HASHCTL transferHashInfo;
+	MemSet(&transferHashInfo, 0, sizeof(HASHCTL));
+	transferHashInfo.keysize = sizeof(NodePair);
+	transferHashInfo.entrysize = sizeof(NodeToNodeFragmentsTransfer);
+	transferHashInfo.hcxt = CurrentMemoryContext;
+	HTAB *transferHash = hash_create("Fragment Transfer Hash", 32, &transferHashInfo,
+									 HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	ListCell *fragmentCell = NULL;
+	foreach(fragmentCell, fragmentList)
+	{
+		DistributedResultFragment *fragment = lfirst(fragmentCell);
+		List *placementList = ShardPlacementList(fragment->targetShardId);
```

Comment:
Would it make sense to use FinalizedShardPlacementList?

Reply:
done